pandas/core/common.py

```diff
@@ -2439,6 +2439,8 @@ def _get_dtype_type(arr_or_dtype):
         return np.dtype(arr_or_dtype).type
     elif isinstance(arr_or_dtype, CategoricalDtype):
         return CategoricalDtypeType
+    elif arr_or_dtype is NotImplemented:
+        return type(NotImplemented)
```

Comment:
OK, thanks for clarifying. My sense is that this method probably *should* raise if it gets `NotImplemented`, because this is not really a valid numpy dtype.

Instead, I would make the wrapper functions in `pandas.core.index` handle `NotImplemented` explicitly, e.g., add something like
```
if result is NotImplemented:
    return result
```

Operations that wrap arithmetic should know how to handle `NotImplemented` -- that's a saner place to put it than here (where it could eat some strange errors)..